rapier/src/pipeline.rs

```diff
@@ -1,19 +1,355 @@
+    /// world
+    ///
+    /// See the [ray_casting](https://github.com/jcornaz/heron/blob/main/examples/ray_casting.rs)
+    /// example for a detailed usage example.
+    #[derive(SystemParam)]
+    pub struct PhysicsWorld<'a> {
+        query_pipeline: ResMut<'a, QueryPipeline>,
+        colliders: ResMut<'a, ColliderSet>,
+    }
+
+    impl<'a> PhysicsWorld<'a> {
+        /// Cast a ray and get the collision shape entity, point, and normal at which it collided,
+        /// if any
+        ///
+        /// - `from`: The point to cast the ray from
+        /// - `to`: The point to cast the ray to
+        /// - `solid`: If `true` a point cast from the inside of a solid object will stop
+        ///   immediately and the collision point will be the same as the `from` point. If `false` a
+        ///   ray cast from inside of an object will act like the object is hollow and will hit the
+        ///   surface of the object after traveling through the object interior.
+        ///
+        /// # Panics
+        ///
+        /// This will panic if the `from` point and the `to` point are the same.
+        #[must_use]
+        pub fn ray_cast(&self, from: Vec3, to: Vec3, solid: bool) -> Option<RayCastInfo> {
+            self.ray_cast_internal(from, to, solid, CollisionLayers::default(), None)
+        }
+
+        /// Cast a ray with extra filters
+        ///
+        /// Behaves the same as [`ray_cast()`][Self::ray_cast] but takes extra arguments for
+        /// filtering results:
+        ///
+        /// - `layers`: The [`CollisionLayers`] to considered for collisions, allowing for coarse
+        ///   filtering of collisions.
+        /// - `filter`: A closure taking an [`Entity`] and returning `true` if the entity should be
+        ///   considered for collisions, allowing for fine-grained, per-entity filtering of
+        ///   collisions.
+        ///
+        /// # Panics
+        ///
+        /// This will panic if the `from` point and the `to` point are the same.
+        #[must_use]
+        pub fn ray_cast_with_filter<F>(
+            &self,
+            from: Vec3,
+            to: Vec3,
+            solid: bool,
+            layers: CollisionLayers,
+            filter: F,
+        ) -> Option<RayCastInfo>
+        where
+            F: Fn(Entity) -> bool,
+        {
+            self.ray_cast_internal(from, to, solid, layers, Some(&filter))
+        }
+
+        /// Non-public implementatino of `ray_cast`
+        #[must_use]
+        #[allow(clippy::cast_possible_truncation)]
+        fn ray_cast_internal(
+            &self,
+            from: Vec3,
+            to: Vec3,
+            solid: bool,
+            layers: CollisionLayers,
+            filter: Option<&dyn Fn(Entity) -> bool>,
+        ) -> Option<RayCastInfo> {
+            assert_ne!(from, to, "From and to points must not be the same");
+            let to_from_diff = to - from;
+            let direction = to_from_diff.normalize();
```

Comment:
Since the result type of the function is `Option`, I'd rather return `None` than panicking.

```suggestion
            let to_from_diff = to - from;
            let direction = match to_from_diff.try_normalize() {
              Some(d) => d,
              None => return None,
            };
```